src/dune_engine/scheduler.ml

```diff
@@ -552,20 +552,27 @@ let ignore_for_watch p =
   let+ t = t () in
   Event.Queue.ignore_next_file_change_event t.events p
 
-exception Cancel_build
-
 let with_job_slot f =
+  let run_only_if_current_status_is_building ~f =
+    let* t = t () in
```

Comment:
If we move the function `run_only_if_current_status_is_building` down by one line, then we won't have to lookup `t ()` three times.

Reply:
Also it can probably change to `raise_if_cancelled : unit -> unit` instead of "wrapping" f.